app/scripts/HorizontalGeneAnnotationsTrack.js

```diff
@@ -187,163 +609,106 @@ class HorizontalGeneAnnotationsTrack extends HorizontalTiled1DPixiTrack {
-              this.drawExons(
-                tile.rectGraphics,
-                txStart,
-                txEnd,
-                exonStarts,
-                exonEnds,
-                chrOffset,
-                yMiddle,
-                geneInfo[5]
-              ).map(x => [x, geneInfo[5]])
-            );
-          }
-        } else {
-          let poly = [];
+    // remove the fillers that are contained within a gene
+    plusFillerRects = plusFillerRects.filter(x => !x.hide);
+    minusFillerRects = minusFillerRects.filter(x => !x.hide);
 
-          if (geneInfo[5] === '+') {
-            poly = [
-              rectX,
-              rectY,
-              rectX + this.geneRectHeight / 2,
-              rectY + this.geneRectHeight / 2,
-              rectX,
-              rectY + this.geneRectHeight
-            ];
-          } else {
-            poly = [
-              rectX,
-              rectY,
-              rectX - this.geneRectHeight / 2,
-              rectY + this.geneRectHeight / 2,
-              rectX,
-              rectY + this.geneRectHeight
-            ];
-          }
-          tile.rectGraphics.drawPolygon(poly);
+    const yMiddle = this.dimensions[1] / 2;
 
-          tile.allRects.push([poly, geneInfo[5]]);
-        }
+    // const fillerGeneSpacing = (this.options.fillerHeight - this.geneRectHeight) / 2;
+    const plusStrandCenterY =
+      yMiddle - this.geneRectHeight / 2 - this.geneStrandSpacing / 2;
+    const minusStrandCenterY =
+      yMiddle + this.geneRectHeight / 2 + this.geneStrandSpacing / 2;
 
-        // tile probably hasn't been initialized yet
-        if (!tile.texts) return;
+    const plusRenderContext = [
+      this,
+      tile,
+      tile.rectGraphics,
+      this._xScale,
+      fill['+'],
+      FILLER_RECT_ALPHA,
+      plusStrandCenterY,
+      this.geneRectHeight
+    ];
+    const minusRenderContext = [
+      this,
+      tile,
+      tile.rectGraphics,
+      this._xScale,
+      fill['-'],
+      FILLER_RECT_ALPHA,
+      minusStrandCenterY,
+      this.geneRectHeight
+    ];
 
-        // don't draw texts for the latter entries in the tile
-        if (i >= MAX_TEXTS) return;
+    renderRects(plusFillerRects, ...plusRenderContext);
+    renderRects(minusFillerRects, ...minusRenderContext);
 
-        const text = tile.texts[geneId];
+    plusRenderContext[5] = GENE_ALPHA;
+    minusRenderContext[5] = GENE_ALPHA;
 
-        if (!text) return;
+    renderGenes(plusGenes, ...plusRenderContext);
+    renderGenes(minusGenes, ...minusRenderContext);
 
-        text.style = {
-          fontSize: `${this.fontSize}px`,
-          fontFamily: FONT_FAMILY,
-          fill: fill[geneInfo[5]]
-        };
+    renderMask(this, tile);
 
-        if (!(geneId in tile.textWidths)) {
-          text.updateTransform();
-          const textWidth = text.getBounds().width;
-          const textHeight = text.getBounds().height;
+    stretchRects(this);
```

Comment:
More of a code style preference, but it seems bad to just have functions where you pass `this` around.

The individual functions do not seem to be using many things from their `track` parameters, it looks like in total in all the functions, the only instance fields that are used are `track._xScale`, `track.renderTile`, `track.tilesetInfo`, `track.fetchedTiles`, `track.dimensions`.

Maybe one place where it is required is `geneClickFunc` where the track needs to be passed to the function

Reply:
In the general case, I agree with you but I feel like in this particular instance there is some logical coherence and elegance to saying `stretchRects(track)` rather than `stretchRects(xScale, renderTile, fetchedTiles)`.

If you don't strongly object, I'll keep it like this and if it becomes troublesome we can refactor in the future.

Reply:
No strong objection, that sounds good to me.